compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeLowering.cpp

```diff
@@ -354,6 +354,413 @@ struct DmaWaitToTctSyncConverter final
+      if (producerLocks.size() != 1) {
+        return targetObjFifo.emitOpError()
+               << "expected a single producer lock for tile: "
+               << channel.getTile();
+      }
+      if (consumerLocks.size() != 1) {
+        return targetObjFifo.emitOpError()
+               << "expected a single consumer lock for tile: "
+               << channel.getTile();
+      }
+      std::pair<AMDAIE::LockOp, AMDAIE::LockOp> lockPair =
+          std::make_pair(producerLocks[0], consumerLocks[0]);
+      SmallVector<int64_t> canonicalizedSizes, canonicalizedStrides;
+      if (failed(foldDimsAndReturnAsStatic(
+              rewriter, deviceModel, targetMixedSizes, targetMixedStrides,
+              canonicalizedSizes, canonicalizedStrides, repetitionCount.value(),
+              targetMemSpace, [&]() { return dmaOp->emitOpError(); }))) {
+        return failure();
+      };
+      if (failed(createDMABlocks(rewriter, tileOp, AMDAIE::DMAChannelDir::S2MM,
+                                 channel.getValue(), canonicalizedSizes,
+                                 canonicalizedStrides, acqNum, acqNum,
+                                 targetOffset, buffers, lockPair, packetId))) {
+        return targetObjFifo.emitOpError() << "could not create DMA operations";
+      }
+    }
+    return success();
+  };
+  std::optional<size_t> maybeOffset = dmaOp.getStaticBaseOffset();
+  if (!maybeOffset) {
+    return dmaOp->emitOpError() << "could not compute a static base offset";
+  }
+  std::optional<uint8_t> maybeMemSpace = dmaOp.getMemorySpaceAsUInt();
+  if (!maybeMemSpace) {
+    return dmaOp->emitOpError() << "expected to have a memory space in input";
+  }
+  if (dmaOp.getInput() == source) {
+    if (failed(processSource(*maybeOffset, *maybeMemSpace,
+                             dmaOp.getMixedSizes(), dmaOp.getMixedStrides()))) {
+      return failure();
+    }
+  } else if (dmaOp.getInput() == target) {
+    if (failed(processTarget(*maybeOffset, *maybeMemSpace,
+                             dmaOp.getMixedSizes(), dmaOp.getMixedStrides()))) {
+      return failure();
+    }
+  } else {
+    return dmaOp->emitOpError()
+           << "expected input of amdaie.npu.half_dma_cpy_nd to be either a "
+              "source or a target of a connection op";
+  }
+  return success();
+}
+
+LogicalResult lowerDmasForReprogramming(Operation *moduleOp,
+                                        AMDAIE::AMDAIEDeviceModel deviceModel) {
+  IRRewriter rewriter(moduleOp->getContext());
+  OpBuilder::InsertionGuard guard(rewriter);
+  Operation *controlCodeOp = nullptr;
+  moduleOp->walk([&](AMDAIE::ControlCodeOp op) { controlCodeOp = op; });
+  DenseMap<Value, Operation *> tileToMemOpMap;
+  int connectionIndex{0};
+  llvm::SmallSetVector<Operation *, 16> toBeErased;
+  auto eraseCandidate = [&](Operation *op) {
+    for (Operation *userOp : op->getUsers()) {
+      toBeErased.insert(userOp);
+    }
+    toBeErased.insert(op);
+  };
```

Comment:
If there is only one use there is no need for a a lambda.

Reply:
Correct. But I prefer doing it this way as it makes the `walk` cycle easier to read.

Reply:
I don't think this is a good use of a lambda because it's only used once and trivial so when reading eraseCandidate below I need to look up what that does, while if it were in line, I would instantly see what's going on.